# context/manager.class.py
class MyContextManager:
	def __init__(self):
		print("MyContextManager init", id(self))

	def __enter__(self):
		print("Entering 'with' context")
		return self

	def __exit__(self, exc_type, exc_val, exc_tb):
		print(f"{exc_type = } {exc_val = } {exc_tb = }")
		print("Exiting 'with' context")
		return True

ctx_mgr = MyContextManager()
print("About to enter 'with' context")
with ctx_mgr as mgr:
	print("Inside the 'with' context")
	print(id(mgr))
	raise Exception("Exception inside 'with' context")
	print("This line will never be reached") # lol, even VS Code hints that it'll never be reached
print("After 'with' context")
